### PR TITLE
Fix pthreads with closure compiler

### PR DESCRIPTION
**Pull request by kripken**
Turns out much of the pthreads runtime was not closure-safe. This PR fixes that, using quoted strings as closure expects.

While doing that I noticed that some of the things in place for pthreads + MODULARIZE would also help closure. In both cases the key thing is that `worker.js`, that loads the main JS file, is separate from it, so it needs proper interfaces and can't just rely on global variables (which in closure are minified in the main JS, and in MODULARIZE are in another scope). So I removed things like `makeAsmExportAccessInPthread, makeAsmGlobalAccessInPthread, makeAsmExportAndGlobalAssignTargetInPthread` which makes the code simpler.

Because of that I was also able to remove a bunch of globals from `worker.js` which further simplifies things, and should also help node.js workers (where the global scope is very different, #6567).

Most of this PR is straightforward according to the above notes. One slightly tricky thing is the stack setup, which I refactored to use a new `Module.applyStackValues` method that `worker.js` calls at the right time (as the stack is set up after the JS loads).

cc @wffurr  @VirtualTim

### CODE REVIEW

**Comment by VirtualTim** [comment]
Had a quick look, and everything looks fine to me. But of course with a change so large it's easy to miss stuff. But all the tests passed, so that makes me a lot more confident.

**Comment by kripken** [comment]
Yeah, I'd like to get even more testing than the test suite if possible, before landing this, as big changes are riskier...

@floooh @brion I think you might have pthreads builds, maybe you can try this PR on them?

**Comment by floooh** [comment]
I haven't been using pthreads with emscripten so far, only the old worker thread approach. Sorry :)

File: src/preamble.js [hunk 1]
```diff
@@ -366,6 +371,24 @@ assert(DYNAMIC_BASE % 16 === 0, 'heap must start aligned');
 #endif
 
 #if USE_PTHREADS
+if (ENVIRONMENT_IS_PTHREAD) {
+  // Pthread workers compute these in worker.js when they run, which is after
+  // this script loads in the worker initially. Set some fake values to
+  // catch bugs, then set the real values at load time.
+  STACK_MAX = STACKTOP = STACK_MAX = 0x7FFFFFFF;
```
Comment (juj) [inline src/preamble.js]: `#if ASSERTIONS`?

**Comment by juj** [comment]
I am not sure how to react to this PR. It is great to be Closure compatible, but being so in a way that requires exporting everything on `Module` is not good on code size (plus it overloads `Module` with another extra meaning). Closure supports minifying multiple scripts at the same time, treating them as being in the same scope, I think that would be usable here to keep the main script and the worker.js script minified in tandem.

**Comment by kripken** [comment]
@juj I agree that there is a code size downside to exporting on Module, and that generally we should avoid it as much as possible.

I think it make sense here for two reasons. First, this code is highly complex, and getting more so: the key thing in our pthreads code is that it is invoked by `worker.js` that was not compiled together with it, and that adds complexity in several build types,

 * MODULARIZE, which you added.
 * Closure, added here.
 * Node.js, will be added I hope, see #6567 (both because users have wanted it for a while, and for us it would be great for testing). See details in that link, but the complexity comes from the Node.js global scope being different than the Web, and there is no true equivalent to `importScripts`.

Exporting on Module makes all those cases much simpler and uniform, we just use the Module object to communicate between the different scopes. That is the normal meaning of Module, its the interface between the optimized code and the outside.

The second reason I think exporting on Module makes sense here is that we only have a small number of very specific things that need that exporting. So the additional code size is quite small. Some numbers on hello world with pthreads and `-O3`:

 * Before this PR: `69,567` bytes.
 * After this PR: `69,644`.
 * After this PR, plus closure: `34,909`.

So this adds only 77 bytes, while also making the code simpler + making it possible to run closure and decrease size by 50% :)

Note that the simplification makes it possible to clean up some unnecessary things, which actually help code size. More is possible there but I didn't want to do too much in a single PR.

File: src/worker.js [hunk 2]
```diff
@@ -118,37 +104,26 @@ this.onmessage = function(e) {
...
 #endif
 
       // Initialize the global "process"-wide fields:
-      {{{ makeAsmExportAndGlobalAssignTargetInPthread('DYNAMIC_BASE') }}} = e.data.DYNAMIC_BASE;
-      {{{ makeAsmExportAndGlobalAssignTargetInPthread('DYNAMICTOP_PTR') }}} = e.data.DYNAMICTOP_PTR;
+      Module['DYNAMIC_BASE'] = e.data.DYNAMIC_BASE;
+      Module['DYNAMICTOP_PTR'] = e.data.DYNAMICTOP_PTR;
 
 #if WASM
-      // The Wasm module will have import fields for STACKTOP and STACK_MAX. At 'load' stage of Worker startup, we are just
```
Comment (dschuff) [inline src/worker.js]: This is a lot of explanatory text we are removing. Is it completely outdated? If not, is there a better place to put it, or an equivalent?

**Comment by dschuff** [comment]
I think the simplifications are worth exporting a few more bytes. Especially if we can make further ones in the future.

File: .circleci/config.yml [hunk 3]
```diff
@@ -239,7 +239,8 @@ commands:
       - run:
           name: download chrome
           command: |
-            wget -O ~/chrome.deb https://dl.google.com/linux/direct/google-chrome-beta_current_amd64.deb
```
[the remark this answers is not in the document]
Comment (kripken) [reply .circleci/config.yml]: Ah yes, thanks, this was a cherry-pick of a temporary workaround. Will remove.

**Pull request merged**